lib/core/src/Cardano/Wallet/Primitive/BlockSummary.hs

```diff
@@ -0,0 +1,200 @@
+--
+data BlockSummary m addr txs = BlockSummary
+    { from  :: !BlockHeader
+    , to    :: !BlockHeader
+    , query :: addr -> m txs
+    } deriving (Generic)
+
+-- | 'BlockSummary' used for light-mode.
+type LightSummary m =
+    BlockSummary m (Either Address RewardAccount) ChainEvents
+
+{-------------------------------------------------------------------------------
+    ChainEvents
+-------------------------------------------------------------------------------}
+-- | 'BlockEvents', orderd by slot.
+newtype ChainEvents = ChainEvents (Map Slot BlockEvents)
+    deriving (Eq, Ord, Show)
+
+mkChainEvents :: Map Slot BlockEvents -> ChainEvents
+mkChainEvents = ChainEvents
+
+instance Semigroup ChainEvents where
+    (ChainEvents bs1) <> (ChainEvents bs2) =
+        ChainEvents $ Map.unionWith mergeSameBlock bs1 bs2
+
+instance Monoid ChainEvents where
+    mempty = ChainEvents mempty
+
+-- | Create 'ChainEvents' from a list of block events
+-- (which do not need to be in order.)
+fromBlockEvents :: [BlockEvents] -> ChainEvents
+fromBlockEvents = ChainEvents
+    . Map.fromListWith mergeSameBlock
+    . map (\bl -> (slot bl, bl))
+    . filter (not . nullBlockEvents)
+
+-- | List of 'BlockEvents', in ascending order.
+-- (No duplicate blocks, all transactions within block in order).
+toAscBlockEvents :: ChainEvents -> [BlockEvents]
+toAscBlockEvents (ChainEvents bs) = Map.elems bs
+
+{-------------------------------------------------------------------------------
+    BlockEvents
+-------------------------------------------------------------------------------}
+-- | Events (such as txs, delegations) within a single block
+-- that are potentially relevant to the wallet.
+-- This can be the entire block, or a pre-filtered version of it.
+data BlockEvents = BlockEvents
+    { slot :: !Slot
+    , blockHeight :: !(Quantity "block" Word32)
+    , transactions :: [(Int, Tx)]
+        -- ^ (Index of the transaction within the block, transaction data)
+        -- INVARIANT: The list is ordered by ascending index.
+    , delegations :: [(Int, DelegationCertificate)]
+        -- ^ (Index of the delegation within the block, delegation data)
+        -- INVARIANT: The list is ordered by ascending index.
+    } deriving (Eq, Ord, Generic, Show)
+
+nullBlockEvents :: BlockEvents -> Bool
+nullBlockEvents BlockEvents{transactions=[],delegations=[]} = True
+nullBlockEvents _ = False
+
+-- | Merge block events that belong to the same block.
+mergeSameBlock :: BlockEvents -> BlockEvents -> BlockEvents
+mergeSameBlock
+    BlockEvents{slot,blockHeight,transactions=txs1,delegations=dlg1}
+    BlockEvents{transactions=txs2,delegations=dlg2}
+  = BlockEvents
+    { slot
+    , blockHeight
+    , transactions = mergeOn fst const txs1 txs2
+    , delegations = mergeOn fst const dlg1 dlg2
+    }
+
+-- | Merge two lists in sorted order. Remove duplicate items.
+mergeOn :: Ord key => (a -> key) -> (a -> a -> a) -> [a] -> [a] -> [a]
```

Comment:
would be good to include some examples here, just to better explain what to expect from `mergeOn`. I wonder if there is no function like that in Data.List or Data.List.Extra...maybe we can use zipWith inside?

Reply:
Fair enough. `mergeOn` is very similar to a merging function that one would use in a mergesort implementation. I'll add more documentation / an example.